### ISSUE TITLE
[FEA] `np.shape` does not work on `cudf.DataFrame`

### ISSUE THREAD

**Issue by mroeschke** [rapidsai/cudf#13045]
Curious if any numpy function that dispatches to `__array_function__` should work on `cudf.DataFrame`

```
In [3]: import numpy as np

In [4]: np.shape(pd.DataFrame([1, 2]))
Out[4]: (2, 1)

In [6]: np.shape(cudf.DataFrame([1, 2]))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[6], line 1
----> 1 np.shape(cudf.DataFrame([1, 2]))

File <__array_function__ internals>:180, in shape(*args, **kwargs)

TypeError: no implementation found for 'numpy.shape' on types that implement __array_function__: [<class 'cudf.core.dataframe.DataFrame'>]
```

**Comment by galipremsagar** [comment]
We currently raise an error for our `DataFrame.__array__` implementation, should we be doing the same for `__array_function__` too? cc: @shwina

**Comment by mroeschke** [comment]
A similar question too for `__array_ufunc__`

```
In [3]: cudf.DataFrame(np.eye(2)) * np.eye(2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 cudf.DataFrame(np.eye(2)) * np.eye(2)

TypeError: operand type(s) all returned NotImplemented from __array_ufunc__(<ufunc 'multiply'>, '__call__',      0    1
0  1.0  0.0
1  0.0  1.0, array([[1., 0.],
       [0., 1.]])): 'DataFrame', 'ndarray'

In [4]: import pandas as pd

In [5]: pd.DataFrame(np.eye(2)) * np.eye(2)
Out[5]: 
     0    1
0  1.0  0.0
1  0.0  1.0
```

**Comment by vyasr** [comment]
No, we should not raise by default. However, we probably want a smarter dispatch pattern here.

The problem with `__array__` is that it provides an implicit conversion pathway from device data to host data. That means that a cudf object could be passed to some other library that accesses this attribute and triggers an implicit conversion to host memory. That's an easy performance footgun to eliminate.

`__array_ufunc__` and `__array_function__` are different since they are dispatch pathways for functions. If possible, we should support them. ufuncs are easier since they are [enumerated by numpy](https://numpy.org/doc/stable/reference/ufuncs.html#available-ufuncs). The set of functions supported by the [`__array_function__` protocol](https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_function__), on the other hand, is not. Ideally, our implementation is designed to try to run the function in a way that we know, and fail if no such way is found, and that is what we want.

The two errors above indicate to me that we may want to clean up our failure pathways, though. I'm not sure what errors look like for a given object that is supposed to support these protocols if they don't support a particular ufunc/array func, but we should endeavor to match that. The easiest way to test is probably to define a trivial array-like object with those attribute and see what the expected traceback is. Alternatively the specs for those protocols should indicate exactly what should be done. We may be returning `NotImplemented` somewhere that we should be doing something else, or accessing `__array_ufunc__` directly without the necessary error checking that numpy does internally when it accesses that attribute.
